File: Source/Src_nd/average.f90
module average_module

  use base_state_geometry_module, only: max_radial_level, finest_radial_level, nr_fine, &
       restrict_base, fill_ghost_base, center, nr_irreg, dr, base_cutoff_density_coord, &
       numdisjointchunks, r_start_coord, r_end_coord
  use amrex_fort_module, only: amrex_spacedim
  use meth_params_module, only: spherical, prob_lo, drdxfac

  implicit none

  private

contains

  subroutine average(lev,lo,hi,phi,p_lo,p_hi,phisum) bind (C,name="average")


    integer         , intent (in   ) :: lev, lo(3), hi(3)
    integer         , intent (in   ) :: p_lo(3), p_hi(3)
    double precision, intent (in   ) :: phi(p_lo(1):p_hi(1),p_lo(2):p_hi(2),p_lo(3):p_hi(3))
    double precision, intent (inout) :: phisum(0:max_radial_level,0:nr_fine-1)

    ! local
    integer          :: j,k

    if (amrex_spacedim .eq. 2) then
       do j=lo(2),hi(2)
          phisum(lev,j) = phisum(lev,j) + sum(phi(lo(1):hi(1),j,0))
       end do
    else if (amrex_spacedim .eq. 3) then
       do k=lo(3),hi(3)
          phisum(lev,k) = phisum(lev,k) + sum(phi(lo(1):hi(1),lo(2):hi(2),k))
       end do
    end if

  end subroutine average

  subroutine divide_phisum_by_ncell(phisum,ncell) bind (C,name="divide_phisum_by_ncell")

    double precision, intent(inout) :: phisum(0:max_radial_level,0:nr_fine-1)
    integer         , intent(in   ) ::  ncell(0:max_radial_level)

    integer :: n,i,r

    ! compute phibar by normalizing phisum
    do n=0,max_radial_level
       do i=1,numdisjointchunks(n)
          do r=r_start_coord(n,i),r_end_coord(n,i)
             phisum(n,r) = phisum(n,r) / dble(ncell(n))
          end do
       end do
    end do

    call restrict_base(phisum,1)
    call fill_ghost_base(phisum,1)

  end subroutine divide_phisum_by_ncell


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  ! spherical subroutines
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

  subroutine average_sphr_irreg(lev,lo,hi,phi,p_lo,p_hi,phisum,ncell, &
       cc_to_r,ccr_lo,ccr_hi) bind (C,name="average_sphr_irreg")


    integer         , intent (in   ) :: lev, lo(3), hi(3)
    integer         , intent (in   ) :: p_lo(3), p_hi(3)
    double precision, intent (in   ) :: phi(p_lo(1):p_hi(1),p_lo(2):p_hi(2),p_lo(3):p_hi(3))
    double precision, intent (inout) :: phisum(0:max_radial_level,0:nr_fine-1)
    integer         , intent (inout) ::  ncell(0:max_radial_level,0:nr_fine-1)
    integer         , intent (in   ) :: ccr_lo(3), ccr_hi(3)
    double precision, intent (in   ) :: cc_to_r(ccr_lo(1):ccr_hi(1), &
         ccr_lo(2):ccr_hi(2),ccr_lo(3):ccr_hi(3))

    ! local
    integer          :: i,j,k,index

    do k=lo(3),hi(3)
       do j=lo(2),hi(2)
          do i=lo(1),hi(1)
             index = cc_to_r(i,j,k)
             phisum(lev,index) = phisum(lev,index) + phi(i,j,k)
             ncell(lev,index) = ncell(lev,index) + 1
          end do
       end do
    end do

  end subroutine average_sphr_irreg

  subroutine divide_phisum_by_ncell_irreg(phisum,ncell) &
       bind (C,name="divide_phisum_by_ncell_irreg")

    double precision, intent(inout) :: phisum(0:max_radial_level,0:nr_fine-1)
    integer         , intent(in   ) ::  ncell(0:max_radial_level,0:nr_fine-1)

    integer :: n,r

    do n=0,max_radial_level
       do r=0,nr_fine-1
          ! divide only if ncell>0
          if (ncell(n,r) > 0) then
             phisum(n,r) = phisum(n,r) / ncell(n,r)
          else
             ! keep value constant if it is outside the cutoff coords
             phisum(n,r) = phisum(n,r-1)
          end if
       end do
    end do

    call restrict_base(phisum,1)
    call fill_ghost_base(phisum,1)

  end subroutine divide_phisum_by_ncell_irreg


  subroutine average_sphr(phisum,phibar,ncell,radii,finest_level) &
       bind (C,name="average_sphr")

    integer         , intent(in   ) :: finest_level
    double precision, intent(inout) :: phisum(0:finest_level,-1:nr_irreg)
    double precision, intent(inout) :: phibar(0:max_radial_level,0:nr_fine-1)
    integer         , intent(inout) ::  ncell(0:finest_level,-1:nr_irreg)
    double precision, intent(inout) :: radii(0:finest_level,-1:nr_irreg+1)

    ! local
    integer          :: j,n,r
    integer          :: min_all, min_lev
    integer          :: max_rcoord(0:finest_level), rcoord(0:finest_level)
    integer          :: stencil_coord(0:finest_level)
    double precision :: radius

    integer          :: which_lev(0:nr_fine-1)

    logical          :: limit

    ! normalize phisum so it actually stores the average at a radius
    do n=0,finest_level
       do r=0,nr_irreg
          if (ncell(n,r) .ne. 0.d0) then
             phisum(n,r) = phisum(n,r) / dble(ncell(n,r))
          end if
       end do
    end do

    ! compute center point for the finest level
    phisum(finest_level,-1) =  (11.d0/8.d0)*phisum(finest_level,0) &
         - (3.d0/8.d0)*phisum(finest_level,1)
    ncell(finest_level,-1) = 1

    ! choose which level to interpolate from
    do n=0,finest_level
       rcoord(n) = 0
    end do

    !$OMP PARALLEL DO PRIVATE(r,radius,n,j,min_all,min_lev) FIRSTPRIVATE(rcoord)
    do r=0,nr_fine-1

       radius = (dble(r)+0.5d0)*dr(0)

       ! for each level, find the closest coordinate
       do n=0,finest_level
          do j=rcoord(n),nr_irreg
             if (abs(radius-radii(n,j)) .lt. abs(radius-radii(n,j+1))) then
                rcoord(n) = j
                exit
             end if
          end do
       end do

       ! make sure closest coordinate is in bounds
       do n=0,finest_level-1
          rcoord(n) = max(rcoord(n),1)
       end do
       do n=0,finest_level
          rcoord(n) = min(rcoord(n),nr_irreg-1)
       end do

       ! choose the level with the largest min over the ncell interpolation points
       which_lev(r) = 0
       min_all = min(ncell(0,rcoord(0)-1), &
            ncell(0,rcoord(0)  ), &
            ncell(0,rcoord(0)+1))

       do n=1,finest_level
          min_lev = min(ncell(n,rcoord(n)-1), &
               ncell(n,rcoord(n)  ), &
               ncell(n,rcoord(n)+1))

          if (min_lev .gt. min_all) then
             min_all = min_lev
             which_lev(r) = n
          end if
       end do

       ! if the min hit count at all levels is zero, we expand the search
       ! to find the closest instance of where the hitcount becomes nonzero
       j = 1
       do while (min_all .eq. 0)
          j = j+1
          do n=0,finest_level
             min_lev = max(ncell(n,max(1,rcoord(n)-j)), &
                  ncell(n,min(rcoord(n)+j,nr_irreg-1)))
             if (min_lev .ne. 0) then
                which_lev(r) = n
                min_all = min_lev
                exit
             end if
          end do
       end do

    end do
    !$OMP END PARALLEL DO

    ! squish the list at each level down to exclude points with no contribution
    do n=0,finest_level
       j=0
       do r=0,nr_irreg
          do while(ncell(n,j) .eq. 0)
             j = j+1
             if (j .gt. nr_irreg) then
                exit
             end if
          end do
          if (j .gt. nr_irreg) then
             phisum(n,r:nr_irreg)   = 1.d99
             radii (n,r:nr_irreg+1) = 1.d99
             max_rcoord(n) = r-1
             exit
          end if
          phisum(n,r) = phisum(n,j)
          radii (n,r) = radii (n,j)
          ncell (n,r) = ncell (n,j)
          j = j+1
          if (j .gt. nr_irreg) then
             max_rcoord(n) = r
             exit
          end if
       end do
    end do

    ! compute phibar
    stencil_coord = 0

    !$OMP PARALLEL DO PRIVATE(r,radius,j,limit) FIRSTPRIVATE(stencil_coord)
    do r=0,nr_fine-1

       radius = (dble(r)+0.5d0)*dr(0)

       ! find the closest coordinate
       do j=stencil_coord(which_lev(r)),max_rcoord(which_lev(r))
          if (abs(radius-radii(which_lev(r) ,j  )) .lt. &
               abs(radius-radii(which_lev(r),j+1))) then
             stencil_coord(which_lev(r)) = j
             exit
          end if
       end do

       ! make sure the interpolation points will be in bounds
       if (which_lev(r) .ne. finest_level) then
          stencil_coord(which_lev(r)) = max(stencil_coord(which_lev(r)),1)
       end if
       stencil_coord(which_lev(r)) = min(stencil_coord(which_lev(r)), &
            max_rcoord(which_lev(r))-1)

       if (r > nr_fine - 1 - drdxfac*2.d0**(finest_level-1)) then
          limit = .false.
       else
          limit = .true.
       end if

       call quad_interp(radius, &
            radii(which_lev(r),stencil_coord(which_lev(r))-1), &
            radii(which_lev(r),stencil_coord(which_lev(r))  ), &
            radii(which_lev(r),stencil_coord(which_lev(r))+1), &
            phibar(0,r), &
            phisum(which_lev(r),stencil_coord(which_lev(r))-1), &
            phisum(which_lev(r),stencil_coord(which_lev(r))  ), &
            phisum(which_lev(r),stencil_coord(which_lev(r))+1), limit)

    end do
    !$OMP END PARALLEL DO

  contains

    subroutine cubic_interp(x,x0,x1,x2,x3,y,y0,y1,y2,y3)

      double precision, intent(in   ) :: x,x0,x1,x2,x3,y0,y1,y2,y3
      double precision, intent(  out) :: y

      y = y0 + (y1-y0)/(x1-x0)*(x-x0) &
           + ((y2-y1)/(x2-x1)-(y1-y0)/(x1-x0))/(x2-x0)*(x-x0)*(x-x1) &
           + ( ((y2-y1)/(x2-x1)-(y1-y0)/(x1-x0))/(x2-x0) &
           -((y3-y2)/(x3-x2)-(y2-y1)/(x2-x1))/(x3-x1) ) / (x3-x0) &
           *(x-x0)*(x-x1)*(x-x2)

      if (y .gt. max(y0,y1,y2,y3)) y = max(y0,y1,y2,y3)
      if (y .lt. min(y0,y1,y2,y3)) y = min(y0,y1,y2,y3)

    end subroutine cubic_interp

    subroutine quad_interp(x,x0,x1,x2,y,y0,y1,y2,limit)

      double precision, intent(in   ) :: x,x0,x1,x2,y0,y1,y2
      double precision, intent(  out) :: y
      logical,          intent(in   ) :: limit

      y = y0 + (y1-y0)/(x1-x0)*(x-x0) &
           + ((y2-y1)/(x2-x1)-(y1-y0)/(x1-x0))/(x2-x0)*(x-x0)*(x-x1)


      if (limit) then
         if (y .gt. max(y0,y1,y2)) y = max(y0,y1,y2)
         if (y .lt. min(y0,y1,y2)) y = min(y0,y1,y2)
      end if

    end subroutine quad_interp

    subroutine lin_interp(x,x0,x1,y,y0,y1)

      double precision, intent(in   ) :: x,x0,x1,y0,y1
      double precision, intent(  out) :: y

      y = y0 + (y1-y0)/(x1-x0)*(x-x0)

    end subroutine lin_interp

  end subroutine average_sphr

  subroutine sum_phi_3d_sphr(lev,lo,hi,phi,p_lo,p_hi,phisum, &
                             radii, finest_level, dx, ncell, &
                             mask,     m_lo, m_hi, use_mask) &
                             bind (C,name="sum_phi_3d_sphr")

    integer         , intent (in   ) :: lev, lo(3), hi(3), finest_level
    integer         , intent (in   ) :: p_lo(3), p_hi(3)
    integer         , intent (in   ) :: m_lo(3), m_hi(3)
    double precision, intent (in   ) :: phi(p_lo(1):p_hi(1),p_lo(2):p_hi(2),p_lo(3):p_hi(3))
    double precision, intent (inout) :: phisum(0:finest_level,-1:nr_irreg)
    double precision, intent (in   ) :: radii(0:finest_level,-1:nr_irreg+1)
    double precision, intent (in   ) :: dx(3)
    integer         , intent (inout) :: ncell(0:finest_level,-1:nr_irreg)
    integer         , intent (in   ) :: mask(m_lo(1):m_hi(1),m_lo(2):m_hi(2),m_lo(3):m_hi(3))
    integer         , intent (in   ) :: use_mask

    ! local
    integer          :: i,j,k,index
    double precision :: x,y,z,radius
    logical          :: cell_valid

    do k=lo(3),hi(3)
       z = prob_lo(3) + (dble(k) + 0.5d0)*dx(3) - center(3)

       do j=lo(2),hi(2)
          y = prob_lo(2) + (dble(j) + 0.5d0)*dx(2) - center(2)

          do i=lo(1),hi(1)
             x = prob_lo(1) + (dble(i) + 0.5d0)*dx(1) - center(1)

             ! make sure the cell isn't covered by finer cells
             cell_valid = .true.
             if ( use_mask .eq. 1 ) then
                if ( (mask(i,j,k).eq.1) ) cell_valid = .false.
             end if

             if (cell_valid) then

                ! compute distance to center
                radius = sqrt(x**2 + y**2 + z**2)

                ! figure out which radii index this point maps into
                index = ((radius / dx(1))**2 - 0.75d0) / 2.d0

                ! due to roundoff error, need to ensure that we are in the proper radial bin
                if (index .lt. nr_irreg) then
                   if (abs(radius-radii(lev,index)) .gt. abs(radius-radii(lev,index+1))) then
                      index = index+1
                   end if
                end if

                phisum(lev,index) = phisum(lev,index) + phi(i,j,k)
                ncell(lev,index)  = ncell(lev,index) + 1

             end if

          end do
       end do
    end do

  end subroutine sum_phi_3d_sphr

  subroutine compute_radii_sphr(lev,radii,finest_level,dx) bind (C,name="compute_radii_sphr")

    integer         , intent(in   ) :: lev, finest_level
    double precision, intent(inout) :: radii(0:finest_level,-1:nr_irreg+1)
    double precision, intent(in   ) :: dx(3)

    integer :: r

    !$OMP PARALLEL DO PRIVATE(r)
    do r=0,nr_irreg
       radii(lev,r) = sqrt(0.75d0+2.d0*r)*dx(1)
    end do
    !$OMP END PARALLEL DO

    radii(lev,nr_irreg+1) = 1.d99
    radii(lev,-1) = 0.d0

  end subroutine compute_radii_sphr

end module average_module

File: Util/postprocess/Average.cpp
#include <Maestro.H>
#include <Postprocess.H>

using namespace amrex;

AMREX_GPU_DEVICE
Real QuadInterp(const Real x, const Real x0, const Real x1, const Real x2,
                const Real y0, const Real y1, const Real y2, bool limit);

// Given a multifab of data (phi), average down to a base state quantity, phibar.
// Assume we are spherical, and the averaging is done at constant radius.

void Postprocess::Average(const Vector<MultiFab>& phi, BaseState<Real>& phibar,
                          int comp) {
    // timer for profiling
    BL_PROFILE_VAR("Postprocessing::Average()", PAverage);

    const auto nr_irreg = base_geom.nr_irreg;

    phibar.setVal(0.0);

    // spherical case with even base state spacing

    // For spherical, we construct a 1D array at each level, phisum, that has space
    // allocated for every possible radius that a cell-center at each level can
    // map into.  The radial locations have been precomputed and stored in radii.
    BaseState<Real> phisum_s(finest_level + 1, nr_irreg + 2);
    auto phisum = phisum_s.array();
    phisum_s.setVal(0.0);
    BaseState<Real> radii_s(finest_level + 1, nr_irreg + 3);
    auto radii = radii_s.array();
    BaseState<int> ncell_s(finest_level + 1, nr_irreg + 2);
    auto ncell = ncell_s.array();
    ncell_s.setVal(0);

    const auto& center_p = center;

    const int fine_lev = finest_level + 1;

    // radii contains every possible distance that a cell-center at the finest
    // level can map into
    for (int lev = 0; lev <= finest_level; ++lev) {
        // Get the index space of the domain
        const auto dx = pgeom[lev].CellSizeArray();

        AMREX_PARALLEL_FOR_1D(nr_irreg + 1, r, {
            radii(lev, r + 1) = std::sqrt(0.75 + 2.0 * Real(r)) * dx[0];
        });
        Gpu::synchronize();

        radii(lev, nr_irreg + 2) = 1.e99;
        radii(lev, 0) = 0.0;
    }

    // loop is over the existing levels (up to finest_level)
    for (int lev = finest_level; lev >= 0; --lev) {
        // Get the grid size of the domain
        const auto dx = pgeom[lev].CellSizeArray();
        const auto prob_lo = pgeom[lev].ProbLoArray();

        // get references to the MultiFabs at level lev
        const MultiFab& phi_mf = phi[lev];

        // create mask assuming refinement ratio = 2
        int finelev = lev + 1;
        if (lev == finest_level) finelev = finest_level;

        const BoxArray& fba = phi[finelev].boxArray();
        const iMultiFab& mask = makeFineMask(phi_mf, fba, IntVect(2));

        // Loop over boxes (make sure mfi takes a cell-centered multifab as an argument)
#ifdef _OPENMP
#pragma omp parallel if (!system::regtest_reduction)
#endif
        for (MFIter mfi(phi_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
            // Get the index space of the valid region
            const Box& tilebox = mfi.tilebox();

            const Array4<const int> mask_arr = mask.array(mfi);
            const Array4<const Real> phi_arr = phi[lev].array(mfi, comp);

            bool use_mask = !(lev == fine_lev - 1);

            AMREX_PARALLEL_FOR_3D(tilebox, i, j, k, {
                Real x = prob_lo[0] + (Real(i) + 0.5) * dx[0] - center_p[0];
                Real y = prob_lo[1] + (Real(j) + 0.5) * dx[1] - center_p[1];
                Real z = prob_lo[2] + (Real(k) + 0.5) * dx[2] - center_p[2];

                // make sure the cell isn't covered by finer cells
                bool cell_valid = true;
                if (use_mask) {
                    if (mask_arr(i, j, k) == 1) cell_valid = false;
                }

                if (cell_valid) {
                    // compute distance to center
                    Real radius = sqrt(x * x + y * y + z * z);

                    // figure out which radii index this point maps into
                    int index = int(round(
                        ((radius / dx[0]) * (radius / dx[0]) - 0.75) / 2.0));

                    // due to roundoff error, need to ensure that we are in the proper radial bin
                    if (index < nr_irreg) {
                        if (fabs(radius - radii(lev, index + 1)) >
                            fabs(radius - radii(lev, index + 2))) {
                            index++;
                        }
                    }

                    amrex::HostDevice::Atomic::Add(&(phisum(lev, index + 1)),
                                                   phi_arr(i, j, k));
                    amrex::HostDevice::Atomic::Add(&(ncell(lev, index + 1)), 1);
                }
            });
        }
    }

    // reduction over boxes to get sum
    ParallelDescriptor::ReduceRealSum(phisum.dataPtr(),
                                      (finest_level + 1) * (nr_irreg + 2));
    ParallelDescriptor::ReduceIntSum(ncell.dataPtr(),
                                     (finest_level + 1) * (nr_irreg + 2));

    // normalize phisum so it actually stores the average at a radius
    for (auto n = 0; n <= finest_level; ++n) {
        for (auto r = 0; r <= nr_irreg; ++r) {
            if (ncell(n, r + 1) != 0) {
                phisum(n, r + 1) /= Real(ncell(n, r + 1));
            }
        }
    }

    BaseState<int> which_lev_s(base_geom.nr_fine);
    auto which_lev = which_lev_s.array();
    BaseState<int> max_rcoord_s(fine_lev);
    auto max_rcoord = max_rcoord_s.array();

    // compute center point for the finest level
    phisum(finest_level, 0) = (11.0 / 8.0) * phisum(finest_level, 1) -
                              (3.0 / 8.0) * phisum(finest_level, 2);
    ncell(finest_level, 0) = 1;

    // choose which level to interpolate from
    const auto dr0 = base_geom.dr(0);
    const auto nrf = base_geom.nr_fine;

    AMREX_PARALLEL_FOR_1D(nrf, r, {
        Real radius = (Real(r) + 0.5) * dr0;
        // Vector<int> rcoord_p(fine_lev, 0);
        int rcoord_p[fine_lev];

        // initialize
        for (int& coord : rcoord_p) {
            coord = 0.0;
        }

        // for each level, find the closest coordinate
        for (auto n = 0; n < fine_lev; ++n) {
            for (auto j = rcoord_p[n]; j <= nr_irreg; ++j) {
                if (fabs(radius - radii(n, j + 1)) <
                    fabs(radius - radii(n, j + 2))) {
                    rcoord_p[n] = j;
                    break;
                }
            }
        }

        // make sure closest coordinate is in bounds
        for (auto n = 0; n < fine_lev - 1; ++n) {
            rcoord_p[n] = amrex::max(rcoord_p[n], 1);
        }
        for (auto n = 0; n < fine_lev; ++n) {
            rcoord_p[n] = amrex::min(rcoord_p[n], nr_irreg - 1);
        }

        // choose the level with the largest min over the ncell interpolation points
        which_lev(r) = 0;

        int min_all =
            amrex::min(ncell(0, rcoord_p[0]), ncell(0, rcoord_p[0] + 1),
                       ncell(0, rcoord_p[0] + 2));

        for (auto n = 1; n < fine_lev; ++n) {
            int min_lev =
                amrex::min(ncell(n, rcoord_p[n]), ncell(n, rcoord_p[n] + 1),
                           ncell(n, rcoord_p[n] + 2));

            if (min_lev > min_all) {
                min_all = min_lev;
                which_lev(r) = n;
            }
        }

        // if the min hit count at all levels is zero, we expand the search
        // to find the closest instance of where the hitcount becomes nonzero
        int j = 1;
        while (min_all == 0) {
            j++;
            for (auto n = 0; n < fine_lev; ++n) {
                int min_lev = amrex::max(
                    ncell(n, amrex::max(1, rcoord_p[n] - j) + 1),
                    ncell(n, amrex::min(rcoord_p[n] + j, nr_irreg - 1) + 1));
                if (min_lev != 0) {
                    which_lev(r) = n;
                    min_all = min_lev;
                    break;
                }
            }
        }
    });
    Gpu::synchronize();

    // squish the list at each level down to exclude points with no contribution
    for (auto n = 0; n <= finest_level; ++n) {
        int j = 0;
        for (auto r = 0; r <= nr_irreg; ++r) {
            while (ncell(n, j + 1) == 0) {
                j++;
                if (j > nr_irreg) {
                    break;
                }
            }
            if (j > nr_irreg) {
                for (auto i = r; i <= nr_irreg; ++i) {
                    phisum(n, i + 1) = 1.e99;
                }
                for (auto i = r; i <= nr_irreg + 1; ++i) {
                    radii(n, i + 1) = 1.e99;
                }
                max_rcoord(n) = r - 1;
                break;
            }
            phisum(n, r + 1) = phisum(n, j + 1);
            radii(n, r + 1) = radii(n, j + 1);
            ncell(n, r + 1) = ncell(n, j + 1);
            j++;
            if (j > nr_irreg) {
                max_rcoord(n) = r;
                break;
            }
        }
    }

    // compute phibar
    const Real drdxfac_loc = drdxfac;
    auto phibar_arr = phibar.array();

    AMREX_PARALLEL_FOR_1D(nrf, r, {
        Real radius = (Real(r) + 0.5) * dr0;
        int stencil_coord = 0;

        // find the closest coordinate
        for (auto j = stencil_coord; j <= max_rcoord(which_lev(r)); ++j) {
            if (fabs(radius - radii(which_lev(r), j + 1)) <
                fabs(radius - radii(which_lev(r), j + 2))) {
                stencil_coord = j;
                break;
            }
        }

        // make sure the interpolation points will be in bounds
        if (which_lev(r) != fine_lev - 1) {
            stencil_coord = amrex::max(stencil_coord, 1);
        }
        stencil_coord = amrex::min(stencil_coord, max_rcoord(which_lev(r)) - 1);

        bool limit = (r <= nrf - 1 - drdxfac_loc * pow(2.0, (fine_lev - 2)));

        phibar_arr(0, r) =
            QuadInterp(radius, radii(which_lev(r), stencil_coord),
                       radii(which_lev(r), stencil_coord + 1),
                       radii(which_lev(r), stencil_coord + 2),
                       phisum(which_lev(r), stencil_coord),
                       phisum(which_lev(r), stencil_coord + 1),
                       phisum(which_lev(r), stencil_coord + 2), limit);
    });
    Gpu::synchronize();
}

AMREX_GPU_DEVICE
Real QuadInterp(const Real x, const Real x0, const Real x1, const Real x2,
                const Real y0, const Real y1, const Real y2, bool limit) {
    Real y = y0 + (y1 - y0) / (x1 - x0) * (x - x0) +
             ((y2 - y1) / (x2 - x1) - (y1 - y0) / (x1 - x0)) / (x2 - x0) *
                 (x - x0) * (x - x1);

    if (limit) {
        if (y > max(y0, max(y1, y2))) y = max(y0, max(y1, y2));
        if (y < min(y0, min(y1, y2))) y = min(y0, min(y1, y2));
    }

    return y;
}

// Given a multifab of data (phi), average down to quantity on a plane, phibar.
// Assume spherical case, and averaging is done in the psi direction,
// resulting in averages on an x-z plane cutting through the north and south poles.

void Postprocess::Average2d(const Vector<MultiFab>& phi,
                            Vector<MultiFab>& phibar, int barcomp,
                            const Vector<Box>& bardomain, int comp) {
    // timer for profiling
    BL_PROFILE_VAR("Postprocessing::Average2d()", PAverage2d);

    // construct a 2D array at finest level for phi to be mapped onto
    Box domain = bardomain[0];
    const int xlen = domain.hiVect()[0] + 1;
    const int ylen = domain.hiVect()[2] + 1;
    // Print() << "HACK: " << xlen << ", " << ylen << std::endl;

    BaseState<Real> phisum_s(xlen, ylen);
    auto phisum = phisum_s.array();
    phisum_s.setVal(0.0);
    BaseState<int> ncell_s(xlen, ylen);
    auto ncell = ncell_s.array();
    ncell_s.setVal(0);

    const auto& center_p = center;

    const auto dxFine = pgeom[finest_level].CellSizeArray();

    // loop is over the existing levels (up to finest_level)
    for (int lev = finest_level; lev >= 0; --lev) {
        // Get the grid size of the domain
        const auto dx = pgeom[lev].CellSizeArray();
        const auto prob_lo = pgeom[lev].ProbLoArray();

        // get references to the MultiFabs at level lev
        const MultiFab& phi_mf = phi[lev];

        // create mask assuming refinement ratio = 2
        int finelev = lev + 1;
        if (lev == finest_level) finelev = finest_level;

        const BoxArray& fba = phi[finelev].boxArray();
        const iMultiFab& mask = makeFineMask(phi_mf, fba, IntVect(2));

        // Loop over boxes (make sure mfi takes a cell-centered multifab as an argument)
#ifdef _OPENMP
#pragma omp parallel if (!system::regtest_reduction)
#endif
        for (MFIter mfi(phi_mf, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
            // Get the index space of the valid region
            const Box& tilebox = mfi.tilebox();

            const Array4<const int> mask_arr = mask.array(mfi);
            const Array4<const Real> phi_arr = phi[lev].array(mfi, comp);

            bool use_mask = !(lev == finest_level);

            AMREX_PARALLEL_FOR_3D(tilebox, i, j, k, {
                Real x = prob_lo[0] + (Real(i) + 0.5) * dx[0] - center_p[0];
                Real y = prob_lo[1] + (Real(j) + 0.5) * dx[1] - center_p[1];
                Real z = prob_lo[2] + (Real(k) + 0.5) * dx[2] - center_p[2];

                // make sure the cell isn't covered by finer cells
                bool cell_valid = true;
                if (use_mask) {
                    if (mask_arr(i, j, k) == 1) cell_valid = false;
                }

                if (cell_valid) {
                    // compute distance to center
                    Real radius = sqrt(x * x + y * y + z * z);

                    // compute distances in mapped coordinates
                    Real xp = sqrt(x * x + y * y);
                    Real yp = center_p[2] + z;

                    // figure out which (i,j) index this point maps into
                    //  TODO: could give smoother results if coarse grid data is
                    //  mapped into multiple fine grids
                    int index_i = int(round(xp / dxFine[0] - 0.5));
                    int index_j = int(round(yp / dxFine[2] - 0.5));

                    // TODO: currently does not take data in domain corners
                    if (index_i < xlen and index_j < ylen) {
                        amrex::HostDevice::Atomic::Add(
                            &(phisum(index_i, index_j)), phi_arr(i, j, k));
                        amrex::HostDevice::Atomic::Add(
                            &(ncell(index_i, index_j)), 1);
                    }
                }
            });
        }
    }

    // reduction over boxes to get sum
    ParallelDescriptor::ReduceRealSum(phisum.dataPtr(), xlen * ylen);
    ParallelDescriptor::ReduceIntSum(ncell.dataPtr(), xlen * ylen);

    // normalize phisum so it actually stores the averages
    for (auto i = 0; i < xlen; ++i) {
        for (auto k = 0; k < ylen; ++k) {
            if (ncell(i, k) != 0) {
                phisum(i, k) /= Real(ncell(i, k));
            }
        }
    }

    // Put average array into 2D MultiFab
    // Loop over boxes (make sure mfi takes a cell-centered multifab as an argument)
#ifdef _OPENMP
#pragma omp parallel if (!system::regtest_reduction)
#endif
    for (MFIter mfi(phibar[0], TilingIfNotGPU()); mfi.isValid(); ++mfi) {
        // Get the index space of the valid region
        const Box& tilebox = mfi.tilebox();
        const Array4<Real> phibar_arr = phibar[0].array(mfi, barcomp);

        AMREX_PARALLEL_FOR_3D(tilebox, i, j, k, {
            phibar_arr(i, j, k) = 0.0;
            if (j == 0) phibar_arr(i, j, k) = phisum(i, k);
        });
    }
}
